# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
df = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
df.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*Replace the text in this markdown cell with your predictions, and be sure to provide justification for your guess.*

## Split the Data into Training and Testing Sets

In [3]:
# Create the labels set `y` and features DataFrame `X`
X = df.copy()
X = X.drop(columns='spam')  # or  X.drop(’y-column-name’, axis=1, inplace=True)

y = df['spam']  # or df[’y-column-name’].values.reshape(-1,1)   y[:5]
# or y = df[‘y-column-name’].ravel()   y[:5]   # prints the values in a straight line
# y = df[’y-column-name’].values


In [4]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
df['spam'].value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [5]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y) 

In [6]:
X

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0               0.00               0.64           0.64           0.0   
1               0.21               0.28           0.50           0.0   
2               0.06               0.00           0.71           0.0   
3               0.00               0.00           0.00           0.0   
4               0.00               0.00           0.00           0.0   
...              ...                ...            ...           ...   
4596            0.31               0.00           0.62           0.0   
4597            0.00               0.00           0.00           0.0   
4598            0.30               0.00           0.30           0.0   
4599            0.96               0.00           0.00           0.0   
4600            0.00               0.00           0.65           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0              0.32            0.00              0.00                0.00   
1              0.14            0.28              0.21                0.07   
2              1.23            0.19              0.19                0.12   
3              0.63            0.00              0.31                0.63   
4              0.63            0.00              0.31                0.63   
...             ...             ...               ...                 ...   
4596           0.00            0.31              0.00                0.00   
4597           0.00            0.00              0.00                0.00   
4598           0.00            0.00              0.00                0.00   
4599           0.32            0.00              0.00                0.00   
4600           0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
0                0.00            0.00  ...                   0.0        0.000   
1                0.00            0.94  ...                   0.0        0.000   
2                0.64            0.25  ...                   0.0        0.010   
3                0.31            0.63  ...                   0.0        0.000   
4                0.31            0.63  ...                   0.0        0.000   
...               ...             ...  ...                   ...          ...   
4596             0.00            0.00  ...                   0.0        0.000   
4597             0.00            0.00  ...                   0.0        0.000   
4598             0.00            0.00  ...                   0.0        0.102   
4599             0.00            0.00  ...                   0.0        0.000   
4600             0.00            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0           0.000          0.0        0.778        0.000        0.000   
1           0.132          0.0        0.372        0.180        0.048   
2           0.143          0.0        0.276        0.184        0.010   
3           0.137          0.0        0.137        0.000        0.000   
4           0.135          0.0        0.135        0.000        0.000   
...           ...          ...          ...          ...          ...   
4596        0.232          0.0        0.000        0.000        0.000   
4597        0.000          0.0        0.353        0.000        0.000   
4598        0.718          0.0        0.000        0.000        0.000   
4599        0.057          0.0        0.000        0.000        0.000   
4600        0.000          0.0        0.125        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
0                          3.756                          61   
1                          5.114                         101   
2                          9.821                         485   
3                          3.537                          40   
4                          3.537              

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [7]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler().fit(X_train)

In [8]:
# Fit the Standard Scaler with the training data
X_train_scaled = scaler.transform(X_train)
X_train_scaled

array([[-0.36038127, -0.16612474, -0.56081367, ..., -0.10729393,
        -0.22985782, -0.43018515],
       [-0.09196258, -0.10723871,  0.56058557, ..., -0.10270527,
         0.05040376,  0.40601422],
       [-0.36038127,  1.20297539, -0.56081367, ..., -0.09582227,
        -0.19248961, -0.32306209],
       ...,
       [-0.05841024, -0.16612474, -0.02962456, ..., -0.0464631 ,
         0.06908786,  0.39482226],
       [-0.36038127, -0.16612474, -0.56081367, ..., -0.10766599,
        -0.20183166, -0.16157808],
       [-0.36038127, -0.16612474, -0.56081367, ..., -0.12983421,
        -0.24854192, -0.43977825]])

In [9]:
# Scale the training data
X_test_scaled = scaler.transform(X_test)
X_test_scaled

array([[-0.36038127, -0.16612474,  0.36384886, ..., -0.12261016,
        -0.23452884, -0.40140582],
       [-0.36038127, -0.16612474, -0.56081367, ..., -0.04264954,
        -0.17847653, -0.36303338],
       [-0.32682893, -0.16612474, -0.50179266, ..., -0.08056805,
        -0.07571395,  3.29513897],
       ...,
       [-0.36038127, -0.16612474,  0.16711215, ..., -0.1290901 ,
        -0.23919987, -0.25591033],
       [ 2.22314863,  0.11358389,  0.95405899, ...,  0.29241613,
         2.19907583,  1.01197896],
       [-0.36038127, -0.16612474,  0.44254355, ..., -0.10664284,
        -0.19248961,  0.01749329]])

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [10]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression
logistic_regression_model = LogisticRegression(random_state=1, max_iter=120) 
#could leave() blank; random_state=?; max_iteration=100 by definition

In [11]:
logistic_regression_model

LogisticRegression(max_iter=120, random_state=1)

In [12]:
logistic_regression_model.fit(X_train_scaled, y_train)
print(f'Training Data Score - train: { logistic_regression_model.score(X_train_scaled, y_train)}')
print(f'Testing Data Score - test: { logistic_regression_model.score(X_test_scaled, y_test)}')

Training Data Score - train: 0.9289855072463769
Testing Data Score - test: 0.9244135534317984


In [19]:
# Make and save testing predictions with the saved logistic regression model using the test data
predictions = logistic_regression_model.predict(X_train_scaled)
results_df = pd.DataFrame({'Prediction': predictions, 'Actual': y_train})

# Review the predictions
predictions

array([0, 0, 0, ..., 1, 0, 0])

In [14]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
testing_predictions = logistic_regression_model.predict(X_test_scaled)

accuracy_score(y_test, testing_predictions)


0.9244135534317984

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [15]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier(random_state=1, n_estimators=500).fit(X_train_scaled, y_train)
print(f'Training Score: {rfc.score(X_train_scaled, y_train)}')
print(f'Training Score: {rfc.score(X_test_scaled, y_test)}')

Training Score: 0.9997101449275362
Training Score: 0.9556907037358818


In [16]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions = rfc.predict(X_test_scaled)

# Review the predictions
predictions

array([0, 0, 0, ..., 1, 0, 0])

In [18]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
acc_score = accuracy_score(y_test, testing_predictions)
print(f'Accuracy Score: {acc_score}')


Accuracy Score: 0.9556907037358818


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

The Random Forest Classifier seems to work best: accuracy score = 0.956 vs 0.924 for Logistic regression